apps/prairielearn/src/lib/groups.sql

```diff
@@ -193,65 +188,50 @@ WITH
     WHERE
       user_id = $user_id
       AND group_id = $group_id
-  ),
-  deleted_group_user_roles AS (
-    DELETE FROM group_user_roles
-    WHERE
-      user_id = $user_id
-      AND group_id = $group_id
   )
 INSERT INTO
   group_logs (authn_user_id, user_id, group_id, action)
 VALUES
   ($authn_user_id, $user_id, $group_id, 'leave');
 
--- BLOCK reassign_group_roles_after_leave
+-- BLOCK update_group_roles
 WITH
-  deleted_group_users_roles AS (
-    DELETE FROM group_user_roles
-    WHERE
-      group_id = $group_id
-  ),
   json_roles AS (
     SELECT
-      (role_assignment ->> 'user_id')::bigint AS user_id,
+      gu.id AS group_user_id,
+      gu.user_id,
+      gu.group_id,
       (role_assignment ->> 'group_role_id')::bigint AS group_role_id
     FROM
       JSON_ARRAY_ELEMENTS($role_assignments::json) AS role_assignment
-  )
-INSERT INTO
-  group_user_roles (group_id, user_id, group_role_id)
-SELECT
-  $group_id AS group_id,
-  user_id,
-  group_role_id
-FROM
-  json_roles
-ON CONFLICT (group_id, user_id, group_role_id) DO
-UPDATE
-SET
-  group_role_id = EXCLUDED.group_role_id;
-
--- BLOCK update_group_roles
-WITH
-  deleted_group_users_roles AS (
-    DELETE FROM group_user_roles
-    WHERE
-      group_id = $group_id
+      JOIN group_users AS gu ON gu.group_id = $group_id
+      AND gu.user_id = (role_assignment ->> 'user_id')::bigint
   ),
   assign_new_group_roles AS (
     INSERT INTO
-      group_user_roles (group_id, user_id, group_role_id)
+      group_user_roles (group_user_id, group_id, user_id, group_role_id)
     SELECT
-      (role_assignment ->> 'group_id')::bigint,
-      (role_assignment ->> 'user_id')::bigint,
-      (role_assignment ->> 'group_role_id')::bigint
+      jr.group_user_id,
+      jr.group_id,
+      jr.user_id,
+      jr.group_role_id
     FROM
-      JSON_ARRAY_ELEMENTS($role_assignments::json) as role_assignment
-    ON CONFLICT (group_id, user_id, group_role_id) DO
-    UPDATE
-    SET
-      group_role_id = EXCLUDED.group_role_id
+      json_roles jr
+    ON CONFLICT DO NOTHING
+  ),
+  deleted_group_users_roles AS (
+    DELETE FROM group_user_roles AS gur
+    WHERE
+      gur.group_id = $group_id
+      AND NOT EXISTS (
+        SELECT
+          1
+        FROM
+          json_roles jr
+        WHERE
+          gur.user_id = jr.user_id
+          AND gur.group_role_id = jr.group_role_id
+      )
   )
 INSERT INTO
   group_logs (authn_user_id, user_id, group_id, action)
```

Comment:
This log entry is inserted with the authorized user instead of the user being affected. I have not changed this behaviour in this PR. I'll leave it for #9106.